### PR TITLE
Representations and Quality Levels

### PR DESCRIPTION
**Pull request by mjneil**
## Description
This PR introduces QualityLevels to contrib-hls.

The first part of this addition is to fix the representations mixin to create closure on the playlist loader instead of the playlist object, since the playlist loader creates a new object reference when loading a new playlist. This would cause outdated references to the representations enabled functions. 

The next change is separating the notion of disabled and blacklisted. We disable playlists for a multitude of reasons. Sometimes we want to blacklist a playlist permanently, for a short time, or to manually have a choice of which playlist is loaded. Since blacklisting and disabling/enabling use the same state variable, there is no way to know the difference between enabling a playlist that was disabled manually or blacklisted permanently.

Lastly, this change brings in videojs-contrib-quality-levels to expose the qualityLevels api on the player and populate the list of quality levels with the available representations in the source. This requires [these changes](https://github.com/videojs/videojs-contrib-quality-levels/pull/4) in contrib-quality-levels.

## Specific Changes proposed
- make representation references persist on new playlist loads
- separate disabling from blacklisting
- bring in the qualityLevels api

## Requirements Checklist
- [x] Feature implemented / Bug fixed
- [x] Unit Tests updated or fixed
- [ ] Docs/guides updated
- [ ] Reviewed by Two Core Contributors

### CODE REVIEW

File: src/videojs-contrib-hls.js [hunk 1]
```diff
@@ -78,6 +79,57 @@ const safeGetComputedStyle = function(el, property) {
...
+ * @param {PlaylistLoader} playlistLoader PlaylistLoader containing the new media info.
+ * @function handleHlsMediaChange
+ */
+const handleHlsMediaChange = function(qualityLevels, playlistLoader) {
+  let newPlaylist = playlistLoader.media();
+  let selectedIndex = getHlsPlaylistIndex(newPlaylist, playlistLoader.master.playlists);
+
+  qualityLevels.selectedIndex_ = selectedIndex;
+  qualityLevels.trigger({
+    selectedIndex,
```
Comment (gesinger) [inline src/videojs-contrib-hls.js]: Is this guaranteed to be the same index as the index in QualityLevelList?

File: src/videojs-contrib-hls.js [hunk 2]
```diff
@@ -524,6 +578,26 @@ class HlsHandler extends Component {
   }
 
   /**
+   * Initializes the quality levels and sets listeners to update them.
+   *
+   * @method setupQualityLevels_
+   * @private
+   */
+  setupQualityLevels_() {
+    let _player = videojs(this.tech_.options_.playerId);
```
Comment (mjneil) [inline src/videojs-contrib-hls.js]: not really, was just following suit of already existing code https://github.com/videojs/videojs-contrib-hls/blob/master/src/videojs-contrib-hls.js#L287

File: src/videojs-contrib-hls.js (hunk 1 above)
Comment (mjneil) [inline src/videojs-contrib-hls.js]: If the list is changes (e.g. a quality level is removed with `removeQualityLevel`, then it wont be the same. We do set the id of the quality level to the same as the index of its representation. With the QualityLevelList api is setup, i could guarantee this in a round about way of finding the qualitylevel by id, then get the index of that level in the underlying list. Or the quality level api could be updated for the getQualityLevelById to instead return an index.

File: test/rendition-mixin.test.js [hunk 3]
```diff
@@ -174,19 +192,20 @@ QUnit.test('setting a representation to disabled sets excludeUntil to Infinity',
 
   renditions[0].enabled(false);
 
-  assert.equal(playlists[0].excludeUntil, Infinity, 'rendition has an infinite excludeUntil');
-  assert.equal(playlists[1].excludeUntil, 0, 'rendition has an excludeUntil of zero');
+  assert.equal(playlists[0].disabled, true, 'rendition has been disabled');
+  assert.equal(playlists[1].disabled, undefined, 'rendition has not been disabled');
```
Comment (mjneil) [inline test/rendition-mixin.test.js]: added (not sure why github didnt pick this up as outdated along with the other comments)

File: src/videojs-contrib-hls.js [hunk 4]
```diff
@@ -106,11 +86,17 @@ const getHlsPlaylistIndex = function(media, playlists) {
  */
 const handleHlsMediaChange = function(qualityLevels, playlistLoader) {
   let newPlaylist = playlistLoader.media();
-  let selectedIndex = getHlsPlaylistIndex(newPlaylist, playlistLoader.master.playlists);
+  let i;
```
Comment (gesinger) [inline src/videojs-contrib-hls.js]: Not a big deal, but I generally like to keep the index vars within the scope of the loop when possible, as outside of it the context is often lost, even when the code is close together. Maybe we can use selectedIndex as the var?

File: src/videojs-contrib-hls.js [hunk 5]
```diff
@@ -106,11 +86,17 @@ const getHlsPlaylistIndex = function(media, playlists) {
...
 
-  qualityLevels.selectedIndex_ = selectedIndex;
+  for (i = qualityLevels.length - 1; i >= 0; i--) {
```
Comment (gesinger) [inline src/videojs-contrib-hls.js]: Any specific reason for going down rather than up?

Comment (mjneil) [inline src/videojs-contrib-hls.js]: Was just trying to avoid an extra check. If no quality level is found with the correct id, i will become -1 at the end of the loop, which is what we want to set selectedIndex if there isnt a selected QualityLevel, but along with your comment above, it definitely makes things less clear and its nothing extra to restructure this in a more clear way

File: src/videojs-contrib-hls.js [hunk 6]
```diff
@@ -78,6 +78,45 @@ const safeGetComputedStyle = function(el, property) {
...
+
+  for (let i = 0; i < qualityLevels.length; i++) {
+    if (qualityLevels[i].id === newPlaylist.uri) {
+      selectedIndex = i;
+      break;
+    }
+  }
+
+  qualityLevels.selectedIndex_ = selectedIndex;
+  qualityLevels.trigger({
```
Comment (gesinger) [inline src/videojs-contrib-hls.js]: Do we want to trigger if -1?

Comment (mjneil) [inline src/videojs-contrib-hls.js]: Yes, although selectedIndex shouldn't ever be -1 unless a user manually changes the quality level list, if listening to the change event, this will allow them to see the invalid state and correct or catch any errors instead of this changing out from under them

File: src/playlist-loader.js [hunk 7]
```diff
@@ -249,8 +251,8 @@ const PlaylistLoader = function(srcUrl, hls, withCredentials) {
     let currentBandwidth = loader.media().attributes.BANDWIDTH || 0;
 
     return !(loader.master.playlists.filter((playlist) => {
-      let enabled = typeof playlist.excludeUntil === 'undefined' ||
-                      playlist.excludeUntil <= Date.now();
```
Comment (imbcmdth) [inline src/playlist-loader.js]: These two lines are the same thing as the lines 233-235. I think making these two lines a function that takes a playlist parameter would be better (something like `isPlaylistEnabled_(playlist) {...}`. Then the `enabledPlaylists_` function becomes `return loader.master.playlists.filter(isPlaylistEnabled_).length` and these two lines become just something like `let enabled = isPlaylistEnabled_(playlist);`

File: src/rendition-mixin.js [hunk 8]
```diff
@@ -11,21 +11,22 @@
  * or if undefined returns the current enabled-state for the playlist
  * @return {Boolean} The current enabled-state of the playlist
  */
-let enableFunction = (playlist, changePlaylistFn, enable) => {
-  let currentlyEnabled = typeof playlist.excludeUntil === 'undefined' ||
-                         playlist.excludeUntil <= Date.now();
+let enableFunction = (loader, playlistUri, changePlaylistFn, enable) => {
+  let playlist = loader.master.playlists[playlistUri];
+  let blacklisted = playlist.excludeUntil && playlist.excludeUntil > Date.now();
+  let currentlyEnabled = (!playlist.disabled && !blacklisted);
```
Comment (imbcmdth) [inline src/rendition-mixin.js]: Another perfect place for a hypothetical `isPlaylistEnabled_` function. 👍

File: src/playlist-loader.js (hunk 7 above)
Comment (mjneil) [inline src/playlist-loader.js]: Good call, I'm thinking an `isEnabled(playlist)` exported from playlist.js since we have other playlist utility functions there

File: src/rendition-mixin.js (hunk 8 above)
Comment (mjneil) [inline src/rendition-mixin.js]: The function is not as useful here since we still need to get the intermediate information of whether the playlist is blacklisted or not. In the other places we only care about if the playlist is enabled or not, but here we want to make sure that the playlist isn't blacklisted even if `enable` differs from `currentlyEnabled` before calling the `changePlaylistFn`

File: src/videojs-contrib-hls.js [hunk 9]
```diff
@@ -103,10 +142,9 @@ Hls.STANDARD_PLAYLIST_SELECTOR = function() {
   // filter out any playlists that have been excluded due to
   // incompatible configurations or playback errors
   sortedPlaylists = sortedPlaylists.filter((localVariant) => {
-    if (typeof localVariant.excludeUntil !== 'undefined') {
-      return now >= localVariant.excludeUntil;
-    }
-    return true;
+    let blacklisted = localVariant.excludeUntil && localVariant.excludeUntil > now;
+
+    return (!localVariant.disabled && !blacklisted);
```
Comment (imbcmdth) [inline src/videojs-contrib-hls.js]: You know what would be really great here!?

File: package.json [hunk 10]
```diff
@@ -96,6 +96,7 @@
     "url-toolkit": "^1.0.4",
     "video.js": "^5.15.1",
     "videojs-contrib-media-sources": "^4.1.3",
+    "videojs-contrib-quality-levels": "^1.0.0",
```
Comment (forbesjo) [inline package.json]: This should be removed

Comment (mjneil) [inline package.json]: Good catch, slipped through the rebase

File: src/rendition-mixin.js [hunk 11]
```diff
@@ -87,7 +92,8 @@ let renditionSelectionMixin = function(hlsHandler) {
     return playlists
       .master
       .playlists
-      .map((e, i) => new Representation(hlsHandler, e, i));
+      .filter((media) => !isBlacklisted(media))
```
Comment (imbcmdth) [inline src/rendition-mixin.js]: I thought we wouldn't filter blacklisted media here?

Comment (mjneil) [inline src/rendition-mixin.js]: So I added this to filter out media that was blacklisted in `excludeIncompatibleVariants_` so that we don't add quality levels for a playlist that will never be selected. It's a good point that since every time representations is called, it rebuilds its list with this filter, so temporary blacklisting will get removed here. With respect to quality levels this isn't an issue since we only call representations once and then work with the quality level list. Could filter on `media.excludeUntil !== Infinity` instead to remove the permanent blacklisted media but still allow temporary blacklists through.

Comment (gesinger) [inline src/rendition-mixin.js]: This may get overly complicated, but would it be better for us to use something else to identify incompatible variants versus temporary suspensions? It may make more contextual sense, but if it ends up being too many options, the check for not Infinity could be an OK solution for now.

File: src/videojs-contrib-hls.js [hunk 12]
```diff
@@ -78,6 +78,45 @@ const safeGetComputedStyle = function(el, property) {
 };
 
 /**
+ * Updates the selcetedIndex of the QualityLevelList when a mediachange happens in hls.
```
Comment (gesinger) [inline src/videojs-contrib-hls.js]: selectedIndex

File: src/videojs-contrib-hls.js (hunk 6 above)
Comment (gesinger) [inline src/videojs-contrib-hls.js]: Makes sense 👍

Comment (gesinger) [inline src/videojs-contrib-hls.js]: Will this trigger a change event even if a playlist refreshes but doesn't change?

File: test/rendition-mixin.test.js [hunk 13]
```diff
@@ -135,35 +149,39 @@ QUnit.test('returns representations with width and height if present', function(
   assert.equal(renditions[2].height, undefined, 'rendition has a height of undefined');
 });
 
-QUnit.test('representations are disabled if their excludeUntil is after Date.now', function(assert) {
+QUnit.test('blacklisted playlists are not included in the representations list', function(assert) {
   let hlsHandler = makeMockHlsHandler([
     {
       bandwidth: 0,
-      excludeUntil: Infinity
+      excludeUntil: Infinity,
```
Comment (gesinger) [inline test/rendition-mixin.test.js]: We should add another with excludeUntil between 0 and Infinity.

File: test/rendition-mixin.test.js [hunk 14]
```diff
@@ -135,35 +149,39 @@ QUnit.test('returns representations with width and height if present', function(
...
     }
   ]);
 
   RenditionMixin(hlsHandler);
 
   let renditions = hlsHandler.representations();
 
-  assert.equal(renditions[0].enabled(), false, 'rendition is not enabled');
-  assert.equal(renditions[1].enabled(), true, 'rendition is enabled');
+  assert.equal(renditions.length, 1, 'rblacklisted rendition not added');
```
Comment (gesinger) [inline test/rendition-mixin.test.js]: blacklisted

File: test/rendition-mixin.test.js (hunk 3 above)
Comment (gesinger) [inline test/rendition-mixin.test.js]: 👍

File: src/rendition-mixin.js [hunk 15]
```diff
@@ -11,21 +12,22 @@
  * or if undefined returns the current enabled-state for the playlist
  * @return {Boolean} The current enabled-state of the playlist
  */
-let enableFunction = (playlist, changePlaylistFn, enable) => {
-  let currentlyEnabled = typeof playlist.excludeUntil === 'undefined' ||
-                         playlist.excludeUntil <= Date.now();
+const enableFunction = (loader, playlistUri, changePlaylistFn, enable) => {
+  const playlist = loader.master.playlists[playlistUri];
```
Comment (mjneil) [inline src/rendition-mixin.js]: Yes, the original function used just the playlist but this PR needed to change that since the call to `bind` below would reference an outdated playlist object since `PlaylistLoader` overwrites the reference when it gets a new playlist. This ensures we are acting on the correct object.

File: src/videojs-contrib-hls.js (hunk 6 above)
Comment (mjneil) [inline src/videojs-contrib-hls.js]: Nope, we fire a `loadedplaylist` event every time we refresh a playlist, but only fire `mediachange` when the playlist uri changes. This function is only called on those `mediachange` events.

File: src/rendition-mixin.js (hunk 11 above)
Comment (mjneil) [inline src/rendition-mixin.js]: I think the current decision is that the current solution is OK for now, and we can revisit it later.

**Comment by mjneil** [comment]
RE: Also would probably be worth adding tests for enableFunction

There are already tests to ensure that calling `representation.enable()` calls the `fastQualityChange_` function and tests to make sure the `fastQualityChange_` function is functioning properly,  so I don't think more tests need to be added

File: test/rendition-mixin.test.js [hunk 16]
```diff
@@ -135,35 +149,44 @@ QUnit.test('returns representations with width and height if present', function(
...
+    },
+    {
+      bandwidth: 0,
+      excludeUntil: 0,
+      uri: 'media1.m3u8'
     },
     {
       bandwidth: 0,
-      excludeUntil: 0
+      excludeUntil: Date.now() + 999999,
```
Comment (gesinger) [inline test/rendition-mixin.test.js]: Minor, but might be useful to have one with expired excludeUntil

Comment (mjneil) [inline test/rendition-mixin.test.js]: That is actually covered with the above playlist `excludeUntil: 0`

Comment (gesinger) [inline test/rendition-mixin.test.js]: That is true, and maybe in this case it would be overdoing it to add another one, but usually it is good to have a non 0 test, as 0 is falsy and often is an edge case.

**Comment by gesinger** [comment]
LGTM
